notifier/stomp/failover.go

```diff
@@ -0,0 +1,83 @@
+package stomp
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"net"
+
+	gostomp "github.com/go-stomp/stomp"
+	"github.com/rs/zerolog"
+)
+
+// FailOver will return the first successful connection made against the provided
+// brokers, or an existing connection if not closed.
+//
+// FailOver is safe for concurrent usage.
+type FailOver struct {
+	Config
+}
+
+// Dial will dial the provided uri in accordance with the
+// provided Config.
+func (f *FailOver) Dial(uri string) (*gostomp.Conn, error) {
+	opts := []func(*gostomp.Conn) error{}
+
+	if f.Login != nil {
+		opts = append(opts, gostomp.ConnOpt.Login(f.Login.Login, f.Login.Passcode))
+	}
+
+	var conn io.ReadWriteCloser
+	var err error
+	if f.tls != nil {
+		conn, err = tls.Dial("tcp", uri, f.tls)
+		if err != nil {
+			return nil, fmt.Errorf("failed to connect to tls broker @ %v: %v", uri, err)
+		}
+	} else {
+		conn, err = net.Dial("tcp", uri)
+		if err != nil {
+			return nil, fmt.Errorf("failed to connect to broker @ %v: %v", uri, err)
+		}
+	}
+
+	stompConn, err := gostomp.Connect(conn, opts...)
+	if err != nil {
+		if conn != nil {
+			conn.Close()
+		}
+		return nil, fmt.Errorf("stomp connect handshake to broker @ %v failed: %v", uri, err)
+	}
+
+	return stompConn, err
+}
+
+// Connection returns a new connection to the first successfully handshook broker.
+//
+// The caller MUST call conn.Disconnect() to close the underlying tcp connection
+// when finished.
+func (f *FailOver) Connection(ctx context.Context) (*gostomp.Conn, error) {
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "notifier/stomp/failover").
+		Logger()
+	ctx = log.WithContext(ctx)
+
+	_, err := f.Config.Validate()
```

Comment:
This doesn't create a tls object in a config since the config is not a pointer

Reply:
what do you mean?

Reply:
Validate() returns a copy of the config, in this case we were only confirming that the config is still valid.